### PR TITLE
[NETBEANS-404] Adjust Javadoc font size to editor's zoom level, and avoid all-monospace font

### PR DESCRIPTION
**Pull request by eirikbakke**

### CODE REVIEW

File: editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java [hunk 1]
```diff
@@ -105,6 +106,16 @@ public DocumentationScrollPane(JTextComponent editorComponent) {
         
         // Add the completion doc view
         view = new HTMLDocView(bgColor);
+        Integer textZoom = (Integer) editorComponent.getClientProperty(TEXT_ZOOM_PROPERTY);
+        // Use the same logic as in o.n.editor.GlyphGutter.update().
```
Comment (cowwoc) [inline editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java]: Shouldn't we use the fully-qualified class name here (no abbreviations)?

Comment (eirikbakke) [reply editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java]: Sure.

(The "o.n." thing seems to be a convention in NetBeans documentation, e.g. see  http://bits.netbeans.org/dev/javadoc/org-netbeans-modules-editor-lib/apichanges.html . But it's not consistently applied everywhere.)

Comment (jlahoda) [reply editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java]: To me, personally, o.n (and o.n.m = org.netbeans.modules) seems quite obvious.

Comment (cowwoc) [reply editor.completion/src/org/netbeans/modules/editor/completion/DocumentationScrollPane.java]: @eirikbakke Go with @jlahoda's preference. He's been working on this codebase much longer than I :) I withdraw my objection.

**Comment by geertjanw** [comment]
Any reason not to merge this?

**Comment by matthiasblaesing** [comment]
I'm a bit mixed. I had a quick view through the codebase. The default style of the HTMLEditorKit is changed/potentially changed at a lot of places:

- autoupdate.ui/src/org/netbeans/modules/autoupdate/ui/wizards/PanelBodyContainer.java
- autoupdate.ui/src/org/netbeans/modules/autoupdate/ui/DetailsPanel.java                                                    
- code.analysis/src/org/netbeans/modules/analysis/ui/AnalysisResultTopComponent.java
- editor.completion/src/org/netbeans/modules/editor/completion/HTMLDocView.java                                                        
- java.j2seplatform/src/org/netbeans/modules/java/j2seplatform/platformdefinition/BrokenPlatformCustomizer.java
- java.navigation/src/org/netbeans/modules/java/navigation/HTMLDocView.java                                                       
- openide.dialogs/src/org/openide/WizardDescriptor.java
- openide.explorer/src/org/openide/explorer/propertysheet/DescriptionComponent.java
- openide.loaders/src/org/openide/loaders/TemplateWizard1.java
- projectui/src/org/netbeans/modules/project/ui/TemplatesPanelGUI.java

all these try to modify the base stylesheet of the HTMLEditorKit, which used for all HTMLEditorKits. Othere places modify the stylesheet of a single document and not for the whole kit:

- subversion/src/org/netbeans/modules/subversion/util/NotifyHtmlPanel.java                                                  
- subversion/src/org/netbeans/modules/subversion/options/SvnOptionsPanel.java
- subversion/src/org/netbeans/modules/subversion/client/MissingClientPanel.java
- openide.awt/src/org/openide/awt/SwingBrowserImpl.java

They either just set base styles of add a filtering layer (SwingBrowserImpl).

My take on this: If we want to change the base styles of the HTMLEditorKit, this should be done by one module and I think the `org.netbeans.swing.plaf` (Look & Feel Customization Library) would be the correct place for that. That way we won't get race-conditions for the customizations. If a module needs an individual styling, the stylesheet can be changed on the document level.

@eirikbakke would you be willing to have a closer look at this? I know that I just turned a small problem into a discussion point, but I think bringing it up is the best option.

If a local option is desired, I would check if the stylesheet of the single document that is used, is reset in `HTMLDocView#setContent`. If the stylesheet stays, I think it would be good to change the document stylesheet. If it is reset, it is IMHO worth checking whether resetting the stylesheet each time is an option.

**Comment by eirikbakke** [comment]
Matthias, let me see if I can manage to understand your comment. Am I correct that this is an objection to the _existing_ use of  "css.addStyleSheet(htmlKit.getStyleSheet())" that is currently present in HTMLDocView.setBodyFontInCSS?

Is there globally shared state here that that I am not aware of? It seems the HTMLEditorKit instance that is modified in "css.addStyleSheet(htmlKit.getStyleSheet())" in HTMLDocView belongs to that HTMLDocView instance only. But there's a comment in HTMLDocView saying "the style sheet is shared by all HTMLEditorKits". Is that the problem here?

Also note that the stylesheet change here is context-sensitive--depending on which editor the documentation popup is opened for, the font size may be different. Should the tweak still, in your opinion, be located in org.netbeans.swing.plaf ?

I probably do not have enough familiarity with the related code to do a large refactoring here.

**Comment by matthiasblaesing** [comment]
Eirik, the culprit is, that the stylesheet you set on an HTMLEditorKit is shared by all instances. The setter codes reads like this:

```java
    public void setStyleSheet(StyleSheet s) {
        if (s == null) {
            AppContext.getAppContext().remove(DEFAULT_STYLES_KEY);
        } else {
            AppContext.getAppContext().put(DEFAULT_STYLES_KEY, s);
        }
    }
```

From the outside it might look as if you are setting an instance attribute, it is for all intents and purposes a global state you are changing. So yes the existing manipulation of the state is IMHO dangerous. Your change is in line with the established code paths - you just made me look at it :-)

You could check if you can apply the css changes to the document stylesheet and not the global css one.

I'm a bit fuzzy how `setFont` and the `addRule` interact, but I tested your changes and they work. So I would be willing to merge the changeset as is and postpone a complete rework for later.

**Comment by eirikbakke** [comment]
> From the outside it might look as if you are setting an instance attribute, it is for all intents and purposes a global state you are changing.
Ah, yes, that was not evident to me when I first wrote that code. Should I create a JIRA ticket for that problem so it can be "postponed-but-not-forgotten"?

**Comment by matthiasblaesing** [comment]
I created https://issues.apache.org/jira/browse/NETBEANS-619 and will merge your change. Thank you for looking into this.

**Comment by eirikbakke** [comment]
Thank you!!

**Pull request merged**